nixos/modules/services/web-apps/lemmy.nix

```diff
@@ -0,0 +1,142 @@
+            description = "Enable Captcha";
+          };
+          difficulty = mkOption {
+            type = types.str; # TODO: Types should be "easy", "medium", "hard". Define an enum.
+            default = "medium";
+            description = "The difficultly of the captcha to solve";
+          };
+        };
+      };
+      default = { };
+      description = "Lemmy configuration";
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.lemmy.settings = {
+      database.user = mkDefault "lemmy";
+      database.host = mkDefault "localhost";
+      database.port = mkDefault 5432;
+      database.database = mkDefault "lemmy";
+      database.pool_size = mkDefault 5;
+
+      bind = mkDefault "127.0.0.1";
+      tls_enabled = mkDefault true;
+      pictrs_url = mkDefault "http://localhost:8080";
+      actor_name_max_length = mkDefault 20;
+
+      rate_limit.message = mkDefault 180;
+      rate_limit.message_per_second = mkDefault 60;
+      rate_limit.post = mkDefault 6;
+      rate_limit.post_per_second = mkDefault 600;
+      rate_limit.register = mkDefault 3;
+      rate_limit.register_per_second = mkDefault 3600;
+      rate_limit.image = mkDefault 6;
+      rate_limit.image_per_second = mkDefault 3600;
+    };
+
+    # TODO: Check if this can't be put under /var/lib
+    environment.etc."lemmy.json".source =
+      settingsFormat.generate "config.hjson" cfg.settings;
+
+    services.postgresql.enable = true;
+
+    systemd.services.lemmy = {
+      environment = {
+        LEMMY_CONFIG_LOCATION = "/etc/lemmy.json";
+
+        # Verify how this is used, and don't put the password in the nix store
+        LEMMY_DATABASE_URL = with services.lemmy.settings.database;"postgres://${user}:${password}@${database}:${port}";
+      };
+
+      documentation = [
+        "https://join-lemmy.org/docs/en/administration/from_scratch.html"
+        "https://join-lemmy.org/docs"
+      ];
+
+      wantedBy = [
+        "multi-user.target"
+      ];
+
+      after = [
+        "lemmy-postgresql.service"
+      ];
+
+      requires = [
+        "lemmy-postgresql.service"
+      ];
+
+      serviceConfig = {
+        # TODO: verify that this is possible
+        DynamicUser = true;
+
+        # TODO: put the configuration here as well
+        StateDirectory = "lemmy";
+
+        ExecStart = "${pkgs.lemmy-server}/bin/lemmy_server";
+      };
+    };
+
+    # TODO: make this configurable
+    systemd.services.lemmy-postgresql = {
+      after = [ "postgresql.service" ];
+      bindsTo = [ "postgresql.service" ];
+      requiredBy = [ "lemmy.service" ];
+      partOf = [ "lemmy.service" ];
+      serviceConfig.Type = "oneshot";
+      unitConfig.ConditionPathExists = "!/var/lib/lemmy/.db-setup";
+      script = ''
+        mkdir -p /var/lib/lemmy/
+        PSQL() {
+          /run/wrappers/bin/sudo -Hu postgres ${config.services.postgresql.package}/bin/psql --port=5432 "$@"
+        }
+        PSQL -tAc "CREATE ROLE lemmy WITH LOGIN;"
```

Comment:
I'm just adding this as a comment so we don't forget, but we should not leave the user and database name hardcoded but take those from the config.